src/audio/selector/selector_generic.c

```diff
@@ -171,171 +171,145 @@ static void sel_s32le_nch(struct comp_dev *dev, struct audio_stream __sparse_cac
 		n = frames - processed;
 		nmax = audio_stream_bytes_without_wrap(source, src) / source_frame_bytes;
 		n = MIN(n, nmax);
-		nmax = audio_stream_bytes_without_wrap(sink, dest) >> BYTES_TO_S16_SAMPLES;
+		nmax = audio_stream_bytes_without_wrap(sink, dst) / sink_frame_bytes;
 		n = MIN(n, nmax);
-		src_ch = src + sel_channel;
-		for (i = 0; i < n; i++) {
-			*dest = *src_ch;
-			src_ch += nch;
-			dest++;
+		for (frm = 0; frm < n; frm++) {
+			process_frame_s16le(dst, sink->channels, src, source->channels,
+					    &cd->coeffs_config);
+			src += source->channels;
+			dst += sink->channels;
 		}
-		src = audio_stream_wrap(source, src + nch * n);
-		dest = audio_stream_wrap(sink, dest);
+		src = audio_stream_wrap(source, src);
+		dst = audio_stream_wrap(sink, dst);
 		processed += n;
 	}
 
-	bsource->consumed += processed >> BYTES_TO_S16_SAMPLES;
-	bsink->size += processed >> BYTES_TO_S16_SAMPLES;
+	bsource->consumed += (processed * source->channels) << BYTES_TO_S16_SAMPLES;
+	bsink->size += (processed * sink->channels) << BYTES_TO_S16_SAMPLES;
 }
+#endif /* CONFIG_FORMAT_S16LE */
 
+#if CONFIG_FORMAT_S24LE || CONFIG_FORMAT_S32LE
 /**
- * \brief Channel selection for 16 bit, at least 2 channels data format.
- * \param[in,out] mod Selector base module device.
- * \param[in,out] bsource Source buffer.
- * \param[in,out] bsink Sink buffer.
- * \param[in] frames Number of frames to process.
+ * \brief Mixing routine for 32-bit, m channel input x n channel output single frame.
+ * \param[out] dst Sink buffer.
+ * \param[in] dst_channels Number of sink channels.
+ * \param[in] src Source data.
+ * \param[in] src_channels Number of source channels.
+ * \param[in] coeffs_config IPC4 micsel config with coefficients.
  */
-static void sel_s16le_nch(struct processing_module *mod, struct input_stream_buffer *bsource,
-			  struct output_stream_buffer *bsink, uint32_t frames)
+static void process_frame_s32le(int32_t dst[], int dst_channels,
+				int32_t src[], int src_channels,
+				struct ipc4_selector_coeffs_config *coeffs_config)
 {
-	struct audio_stream __sparse_cache *source = bsource->data;
-	struct audio_stream __sparse_cache *sink = bsink->data;
-	int8_t *src = (int8_t *)source->r_ptr;
-	int8_t *dst = (int8_t *)sink->w_ptr;
-	int bmax;
-	int b;
-	int bytes_copied = 0;
-	const int bytes_total = frames * audio_stream_frame_bytes(source);
+	int64_t accum;
+	int i;
+	int j;
 
-	while (bytes_copied < bytes_total) {
-		b = bytes_total - bytes_copied;
-		bmax = audio_stream_bytes_without_wrap(source, src);
-		b = MIN(b, bmax);
-		bmax = audio_stream_bytes_without_wrap(sink, dst);
-		b = MIN(b, bmax);
-		memcpy_s(dst, b, src, b);
-		src = audio_stream_wrap(source, src + b);
-		dst = audio_stream_wrap(sink, dst + b);
-		bytes_copied += b;
+	for (i = 0; i < dst_channels; i++) {
+		accum = 0;
+		for (j = 0; j < src_channels; j++)
+			accum += (int64_t)src[j] * (int64_t)coeffs_config->coeffs[i][j];
+		dst[i] = (int32_t)(accum >> 10);
 	}
```

Comment:
isn't this the same multiplication as in `process_frame_s16le()`? If yes - reuse?

Reply:
Functions 'process_frame_s32le' and 'process_frame_s16le', although similar, are significantly different. Basically, one performs 32-bit x 16-bit multiplications and 64-bit additions, while the other performs 16-bit x 16-bit multiplications and 32-bit additions. For this reason I would stay with two separate ones.